app/renderer/components/tabs/tab.js

```diff
@@ -189,34 +189,39 @@ class Tab extends React.Component {
     }
   }
 
-  get tabSize () {
-    const tab = this.tabNode
-    // Avoid TypeError keeping it null until component is mounted
-    return tab && !this.props.isPinnedTab ? tab.getBoundingClientRect().width : null
-  }
-
-  onUpdateTabSize () {
-    const currentSize = getTabBreakpoint(this.tabSize)
-    // Avoid updating breakpoint when user enters fullscreen (see #7301)
-    // Also there can be a race condition for pinned tabs if we update when not needed
-    // since a new tab component with the same key gets created which is not pinned.
-    if (this.props.breakpoint !== currentSize && !this.props.hasTabInFullScreen) {
-      windowActions.setTabBreakpoint(this.props.frameKey, currentSize)
-    }
-  }
-
   componentDidMount () {
-    this.onUpdateTabSize()
+    if (this.props.isPinned) {
+      this.observer.unobserve(this.tabSentinel)
+    }
+    const threshold = Object.values(globalStyles.intersection)
+    // At this moment Chrome can't handle unitless zeroes for rootMargin
+    // see https://github.com/w3c/IntersectionObserver/issues/244
+    const margin = '0px'
     this.tabNode.addEventListener('auxclick', this.onAuxClick.bind(this))
-    window.addEventListener('resize', throttle(this.onUpdateTabSize, tabUpdateFrameRate), { passive: true })
+
+    this.observer = setObserver(this.tabSentinel, threshold, margin, this.onObserve)
+    this.observer.observe(this.tabSentinel)
   }
 
   componentDidUpdate () {
-    this.onUpdateTabSize()
+    // if a tab was just pinned, unobserve its intersection
+    if (this.props.isPinnedTab) {
+      this.observer.unobserve(this.tabSentinel)
+    }
   }
 
   componentWillUnmount () {
-    window.removeEventListener('resize', this.onUpdateTabSize)
+    this.observer.unobserve(this.tabSentinel)
+  }
+
+  onObserve (entries) {
+    // avoid observing pinned tabs
+    if (this.props.isPinnedTab) {
+      return
+    }
+    // we only have one entry
+    const entry = entries[0]
+    windowActions.setTabIntersectionState(this.props.frameKey, entry.intersectionRatio)
```

Comment:
This is called quite a lot of time when resizing, can we doubnce it?

Reply:
throttling tabs as we do now leads to #7925. I tried minimizing perf cost by observing only the first tab in a tab set. I'll share perf wins once I finish this monster